### PR TITLE
i#5843 scheduler: Refactor readers to be single-input

### PR DESCRIPTION
**Pull request by derekbruening**
Removes multi-input support from file_reader_t and other readers now that the scheduler_t owns that.  Specifically:

+ Removes read_next_thread_entry() and requires that read_next_entry() always check the queue (via a provided helper function).

+ Removes skip_thread_instructions() and refactors the pre-skip header reading and the post-skip walking while remembering timestamps. Places these latter two inside reader_t for use by all readers, with zipfile overriding just the fast skip in the middle and sharing all the other code.  This refactoring and sharing solves the problem of missing timestamps when skipping from the middle.

+ Removes the arrays of data for multiple inputs from file_reader_t and all subclasses.

Updates the view_test to use a scheduler for its multiple-input mock reader.

While at it, removes is_complete().

Issue: #5843, #5538

### CODE REVIEW

File: clients/drcachesim/reader/reader.h [hunk 1]
```diff
@@ -182,24 +189,21 @@ class reader_t : public std::iterator<std::input_iterator_tag, memref_t>,
...
     virtual trace_entry_t *
     read_next_entry() = 0;
-    // This first checks the local queue before calling read_next_entry().
-    // in timestamp order.
+    // Returns and removes the entry (nullptr if none) from the local queue.
+    // This should be called by read_next_entry() prior to reading a new record
+    // from the input stream.
     virtual trace_entry_t *
-    read_next_entry_or_queue();
+    read_queue();
```
Comment (abhinav92003) [inline clients/drcachesim/reader/reader.h:202]: s/read_queue/read_buffered_entry/

Or some other name that makes it clear what the queue is.

File: clients/drcachesim/reader/file_reader.h [hunk 2]
```diff
@@ -107,287 +95,83 @@ template <typename T> class file_reader_t : public reader_t {
...
+        // early marker with the arch tag, and version 2 which only differs wrt
+        // TRACE_MARKER_TYPE_KERNEL_EVENT..
+        if (entry->addr > TRACE_ENTRY_VERSION) {
+            ERRMSG("Cannot handle version #%zu (expect version <= #%u)\n", entry->addr,
+                   TRACE_ENTRY_VERSION);
+            return false;
+        }
+        // Read the meta entries until we hit the pid.
+        // We want to pass the tid+pid to the reader *before* any markers,
+        // even though 2 markers preced the tid+pid in the file.
```
Comment (abhinav92003) [inline clients/drcachesim/reader/file_reader.h:134]: s/preced/precede/

File: clients/drcachesim/reader/file_reader.h [hunk 3]
```diff
@@ -107,287 +95,83 @@ template <typename T> class file_reader_t : public reader_t {
...
+        std::queue<trace_entry_t> marker_queue;
+        while ((entry = read_next_entry()) != nullptr) {
+            if (entry->type == TRACE_TYPE_PID) {
+                // We assume the pid entry is the last, right before the timestamp.
+                pid = *entry;
+                break;
+            } else if (entry->type == TRACE_TYPE_THREAD)
+                tid = *entry;
+            else if (entry->type == TRACE_TYPE_MARKER)
+                marker_queue.push(*entry);
```
Comment (abhinav92003) [inline clients/drcachesim/reader/file_reader.h:144]: Add to comment which markers we expect here. Looking at `offline_instr_t::append_thread_header` I see only the OFFLINE_EXT_TYPE_HEADER entry before the tid one but no marker.

File: clients/drcachesim/reader/reader.h [hunk 4]
```diff
@@ -110,18 +110,25 @@ class reader_t : public std::iterator<std::input_iterator_tag, memref_t>,
...
+    // timestamp (plus cpuid) is skipped, the most recent skipped timestamp will be
+    // duplicated prior to the target instruction.  Top-level headers will still be
+    // observed.  This generally should call pre_skip_instructions() to observe the
+    // headers, perform any fast skipping, and then should call
+    // skip_instructions_with_timestamp() to properly duplicate the prior timestamp.
     virtual reader_t &
     skip_instructions(uint64_t instruction_count);
 
+    virtual void
+    pre_skip_instructions();
```
Comment (abhinav92003) [inline clients/drcachesim/reader/reader.h:123]: Why are pre_skip_instructions and skip_instructions_with_timestamp public?

**Review by abhinav92003 — APPROVED** [review]

File: clients/drcachesim/reader/file_reader.h (hunk 2 above)
Comment (derekbruening) [reply clients/drcachesim/reader/file_reader.h:134]: Done.

File: clients/drcachesim/reader/file_reader.h (hunk 3 above)
Comment (derekbruening) [reply clients/drcachesim/reader/file_reader.h:144]: It is legacy traces which have something different: I think they used to have the timestamp before the tid+pid in the raw file.  Updated the comment.

File: clients/drcachesim/reader/reader.h (hunk 4 above)
Comment (derekbruening) [reply clients/drcachesim/reader/reader.h:123]: Shouldn't be: moved to protected.

**Pull request merged**